sql/src/main/java/org/apache/druid/sql/calcite/rule/DruidJoinRule.java

```diff
@@ -222,164 +233,53 @@ private static RexNode makeNullableIfLiteral(final RexNode rexNode, final RexBui
-        plannerContext.setPlanningError(
-            "SQL requires a join with '%s' condition that is not supported.",
-            subCondition.getKind()
-        );
-        return Optional.empty();
-      }
-
-      if (isLeftExpression(firstOperand, numLeftFields) && isRightInputRef(secondOperand, numLeftFields)) {
-        equalitySubConditions.add(new RexEquality(firstOperand, (RexInputRef) secondOperand, comparisonKind));
-        rightColumns.add((RexInputRef) secondOperand);
-      } else if (isRightInputRef(firstOperand, numLeftFields)
-                 && isLeftExpression(secondOperand, numLeftFields)) {
-        equalitySubConditions.add(new RexEquality(secondOperand, (RexInputRef) firstOperand, subCondition.getKind()));
-        rightColumns.add((RexInputRef) firstOperand);
-      } else {
-        // Cannot handle this condition.
-        plannerContext.setPlanningError("SQL is resulting in a join that has unsupported operand types.");
-        return Optional.empty();
-      }
-    }
-
+    ConditionAnalysis conditionAnalysis = analyzeCondition(condition, leftRowType, rexBuilder);
     // if the right side requires a subquery, then even lookup will be transformed to a QueryDataSource
     // thereby allowing join conditions on both k and v columns of the lookup
     if (right != null
         && !DruidJoinQueryRel.computeRightRequiresSubquery(plannerContext, DruidJoinQueryRel.getSomeDruidChild(right))
         && right instanceof DruidQueryRel) {
       DruidQueryRel druidQueryRel = (DruidQueryRel) right;
       if (druidQueryRel.getDruidTable().getDataSource() instanceof LookupDataSource) {
-        long distinctRightColumns = rightColumns.stream().map(RexSlot::getIndex).distinct().count();
+        long distinctRightColumns = conditionAnalysis.rightColumns.stream().map(RexSlot::getIndex).distinct().count();
         if (distinctRightColumns > 1) {
           // it means that the join's right side is lookup and the join condition contains both key and value columns of lookup.
           // currently, the lookup datasource in the native engine doesn't support using value column in the join condition.
           plannerContext.setPlanningError(
               "SQL is resulting in a join involving lookup where value column is used in the condition.");
-          return Optional.empty();
+          return false;
         }
       }
     }
 
-    return Optional.of(
-        new ConditionAnalysis(
-            numLeftFields,
-            equalitySubConditions,
-            literalSubConditions
-        )
-    );
-  }
+    if (joinType != JoinRelType.INNER || !systemFieldList.isEmpty() || NullHandling.replaceWithDefault()) {
+      // I am not sure in what case, the list of system fields will be not empty. I have just picked up this logic
+      // directly from https://github.com/apache/calcite/blob/calcite-1.35.0/core/src/main/java/org/apache/calcite/rel/rules/AbstractJoinExtractFilterRule.java#L58
```

Comment:
what is systemFieldList? i dug around a bit and isnt super obvious to me 🙃 

```
   * @param systemFieldList  List of system fields that will be prefixed to
   *                         output row type; typically empty but must not be
   *                         null
 ``` 
or `Returns a list of system fields that will be prefixed to output row type.` are all that i can find but i don't quite fully understand what it means

Reply:
I am not sure either. I had asked @kgyrtkirk, and he didn't have any answer either. Maybe @gianm knows since he was exchanging emails with calcite community recently over other kind of system fields (like file names in input format). 
In any case, I don't think this list will ever be non-empty based on its usages I found in calcite.